File: sorting/inplace_insertion_sort/insertion_sort.cpp
#include <array>
#include <utility>

template <typename T>
void insertion_sort(T *elements, std::size_t size) {
    if (elements == nullptr) {
        return;
    }

    for (auto i = 1ul; i < size; i++) {
        auto cur_index = i;

        for (auto j = 0ul; j < i; j++) {
            auto cmp_index = i - 1 - j;

            if (elements[cur_index] < elements[cmp_index]) {
                std::swap(elements[cmp_index], elements[cur_index]);
                cur_index--;
            }
        }
    }
}

File: sorting/inplace_insertion_sort/insertion_sort.js
'use strict';

function swap(elements, index0, index1) {
	const tmp = elements[index0];
	elements[index0] = elements[index1];
	elements[index1] = tmp;
}

function insertionSort(elements) {
	for (let i = 1; i < elements.length; i++) {
		for (let j = 0; j < i; j++) {
			if (elements[i] < elements[j]) {
				swap(elements, i, j);
			}
		}
	}
}

